tasks/appraisal.rake

```diff
@@ -0,0 +1,187 @@
+
+namespace :appraisal do # rubocop:disable Metrics/BlockLength
+  def ruby_versions(versions)
+    return TRACER_VERSIONS if versions.empty?
+
+    TRACER_VERSIONS & versions
+  end
+
+  def bundler_version(ruby_version)
+    FORCE_BUNDLER_VERSION[ruby_version]
+  end
+
+  def force_bundler_version(ruby_version)
+    force_gem_version(bundler_version(ruby_version))
+  end
+
+  def force_gem_version(version)
+    return if version.nil?
+
+    "_#{version}_"
+  end
+
+  def bundle(ruby_version)
+    ['bundle', force_bundler_version(ruby_version)].compact
+  end
+
+  def docker(ruby_version, cmd)
+    [
+      'docker-compose', 'run',
+      '--no-deps',                                   # don't start services
+      '-e', 'APPRAISAL_GROUP',                       # pass appraisal group if defined
+      '-e', 'APPRAISAL_SKIP_BUNDLE_CHECK',           # pass appraisal check skip if defined
+      '--rm',                                        # clean up container
+      "tracer-#{ruby_version}",                      # use specific ruby engine and version
+      '/bin/bash', '-c', "'#{cmd.join(' ')}'"        # call command in bash
+    ]
+  end
+
+  def lockfile_prefix(ruby_version)
+    ruby_version = {
+      '2.1' => '2.1.10',
+      '2.2' => '2.2.10',
+      '2.3' => '2.3.8',
+      '2.4' => '2.4.10',
+      '2.5' => '2.5.9',
+      '2.6' => '2.6.10',
+      '2.7' => '2.7.6',
+      '3.0' => '3.0.4',
+      '3.1' => '3.1.2',
+      '3.2' => '3.2.0',
+      'jruby-9.2.8.0' => 'jruby-9.2.8.0',
+      'jruby-9.2' => 'jruby-9.2.21.0',
+      'jruby-9.3' => 'jruby-9.3.9.0',
+      'jruby-9.4' => 'jruby-9.4.0.0',
+    }[ruby_version]
+
+    return "ruby_#{ruby_version}" if ruby_version =~ /^\d/
+
+    ruby_version.tr('-', '_')
+  end
+
+  desc 'Generate Appraisal gemfiles. Takes a version list as argument, defaults to all'
+  task :generate do |_task, args|
+    ruby_versions(args.to_a).each do |ruby_version|
+      cmd = []
+      cmd << ['rm', '-vf', 'Gemfile.lock']
+      cmd << ['gem', 'install', 'bundler', '-v', bundler_version(ruby_version)] if bundler_version(ruby_version)
+      cmd << [*bundle(ruby_version), 'config', 'without', 'check']
+      cmd << [*bundle(ruby_version), 'install']
+      cmd << [*bundle(ruby_version), 'exec', 'appraisal', 'generate']
+      cmd << ['rm', '-vf', 'Gemfile.lock']
+
+      cmd = cmd.map { |c| c << '&&' }.flatten.tap(&:pop)
+
+      p cmd
+      p docker(ruby_version, cmd)
+      sh docker(ruby_version, cmd).join(' ')
+    end
+  end
+
+  desc 'Install Appraisal gemfiles. Takes a version list as argument, defaults to all'
+  task :install do |_task, args|
+    ENV['APPRAISAL_SKIP_BUNDLE_CHECK'] = 'y'
+
+    ruby_versions(args.to_a).each do |ruby_version|
+      cmd = []
+      cmd << ['rm', '-vf', 'Gemfile.lock']
+      cmd << ['gem', 'install', 'bundler', '-v', bundler_version(ruby_version)] if bundler_version(ruby_version)
+      cmd << [*bundle(ruby_version), 'config', 'without', 'check']
```

Comment:
They are unrelated.

`bundle config without check` is about skipping the `:check` group in the main `Gemfile`.

`APPRAISAL_SKIP_BUNDLE_CHECK` is about `appraisal install` running `bundle check` before `bundle install`, because `bundle check` kind of needs the lockfile to exist or things blow up in flames.

Now for the nasty bits (strap yourself tight, it takes several wetware round trips to have that one percolate):

- passing `--without blah` to `appraisal install` is designed to pass `--without blah` to `bundle install`
- passing `--without blah` to `appraisal install` also does not add `bundle check` to the shell command (because reasons? like `bundle check` does not accept `--without`? TBH not sure)
- it turns out that to achieve not running `bundle check`, the only way without monkeypatching the appraisal gem is to pass `--without`: there is no `--skip-bundle-check`. This could be an upstream change but it woudl not be available on older `appraisal` versions, which we need on older rubies
- in addition, the `--without` bundler flag is deprecated for good reasons (because it's stateful+side-effectful but doesn't look like so): it actually works by changing the config
- it turns out that `appraisal` could care less about whatever `without` is in the config, and so setting the config would not disable `bundle check`
- so I resorted to some gnarly monkeypatching (replacing `bundle check` by `bundle exec false`)
- completely unrelatedly to the above, it also turns out that now we want to ignore Gemfile groups such as `:check`, so `bundle config without check` is there but has no relationship whatsoever with `appraisal`.

There are really two parts to the `appraisal` gem: one is to assemble and generate the gemfiles, and the other is a glorified `env BUNDLE_GEMFILE=gemfiles/foo.gemfile bundle something` command wrapper that is not even remotely consistent with the `bundler` CLI (well, IIRC it is, but with *very* early versions back when there was no `bundle exec`). After all this investigation, fixing, and peeking at the `appraisal` source code, the former is okay-ish and does add some value to us, but the latter should IMHO be ripped straight out and we should just call `env BUNDLE_GEMFILE=gemfiles/foo.gemfile bundle something` (with `gemfiles/foo.gemfile` which should be provided by an `appraisal` command which sadly does not exist) instead of trying to beat `appraisal` into submission. But I kinda felt that was out of the scope of the PR ;)

Reply:
Thanks for the detailed explanation!

I really like the idea of calling bundler directly to run the tests instead of using appraisal, leaving appraisal just the task of effectively being a harness to generate gemfiles.